### PR TITLE
Add support for vmcx (Hyper-V binary config format) to the Hyper-V import script

### PR DESCRIPTION
**Pull request by bbrala**
Added an import script that reads the new vmcx format with powershell. It uses native commands to read the format and create the machine the same way as it does right now with the XML's. This way there is no need anymore to hack an XML to get Hyper-V boxes to work.

As Hyper-V support seems to get close in packer, we really need this small feature added so we can start using those boxes as quickly as possible.

This should fix: #7706 and #6102

### CODE REVIEW

**Comment by bbrala** [comment]
Can you provide the vagrant file? I've tried a few and didn't run into problems.

**Comment by irab** [comment]
My vagrant file is super minimal, just:

Vagrant.configure("2") do |config|
  config.vm.box = "ira/leap"
end

File: plugins/providers/hyperv/scripts/import_vm_vmcx.ps1 [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    [string]$auto_start_action=$null,
+    [string]$auto_stop_action=$null
+)
+
+# Include the following modules
+$Dir = Split-Path $script:MyInvocation.MyCommand.Path
+. ([System.IO.Path]::Combine($Dir, "utils\write_messages.ps1"))
+
+# load the config from the vmcx and make a copy for editing, use TMP path so we are sure there is no vhd at the destination
+$vmConfig = (Compare-VM -Copy -Path $vm_config -GenerateNewID -VhdDestinationPath $env:Temp)
```
Comment (bbrala) [inline plugins/providers/hyperv/scripts/import_vm_vmcx.ps1]: Ok, I might've did an oops with the laat commit there. I'll check when I have a pc available.

**Comment by irab** [comment]
There also seems to be another weird problem but it is probably to do with my linux vagrant box:

Using "vagrant up" to start a this Debian Jessie VM creates a VM but it can't boot from HDD due to "no x64-based uefi boot loader was found"

But if I import the VM from the downloaded box files using the Hyper-V GUI it boots fine... probably Vagrant default import settings messing with the import.

**Comment by bbrala** [comment]
I will investigate that box, see what is missing in the import.

**Comment by bbrala** [comment]
Ok, i've looked into it, and it seems it is missing a part of the firmware (what used to be the bios). It should boot from an EFI partition, but that is not imported. I'll have a look how one would go about importing that.

**Comment by bbrala** [comment]
Well, that is fun. I can't find a way to manually set an EFI file to boot from. The file is contained in the firmware, but since we create a whole new VM, it is not present there. Will try again soon to see if i missed something, i'm kind of tired right now. But if not, i might need to use a real copy to get the VM up and running, which should be doable, but required a bit of a different approach.

**Comment by irab** [comment]
Looks like there are [issues with importing any Gen 2 Linux VHDs](https://blogs.msdn.microsoft.com/virtual_pc_guy/2015/02/11/copying-the-vhd-of-a-generation-2-linux-vmand-not-booting-afterwards/). 

Tried the 'fix' below on original VM

grub-install --target=x86_64-efi --efi-directory=/boot/efi --no-nvram --removable

Recreated a box file, re-did "vagrant up --provider hyperv" and VM now boots perfectly. Thanks for the getting the import working!!

**Comment by bbrala** [comment]
Hmm, that issue is because of recreating the whole VM instead of making a copy. Although recreating is how it works right now, i'm not entirely sure it is the best way.

Pros would be the fact you have more control on what features are enabled/set, if we would be copying the VM you never really know what features are enabled. This might make support a bit harder since you have less control.

For now, glad it works. :)

**Comment by bbrala** [comment]
There is already a mention of this problem in the docs it seems, so no real loss of functionality if it doesn't support EPI in firmware.

https://www.vagrantup.com/docs/hyperv/boxes.html

**Comment by bbrala** [comment]
For testing I made a gist explaining how to get vagrant working from source on Windows.

https://gist.github.com/bbrala/7e96be8a4e26f14834e8bbc21d246c38

Just checkout the branch in the repository after running the commands in the gist.

**Comment by bbrala** [comment]
Ok, couldn't resist, the whole New-VM seems flaky sometimes so I used the proper PowerShell command `Compare-VM`  to create the new VM.

- Implemented the differencing disk for vmcx.
- Disk is now copied by Hyper-V instead of Ruby for machines with the new config.
- EFI Firmware now works for machines since it is a real copy. Compare-VM will report incompatibilities should they be found, but i didn't really find anything that could be incompatible with a host.

Only thing i'm slightly worried about is how Windows 8 Pro will perform. But i feel this PR is done.

**Comment by bbrala** [comment]
@chrisroberts it would be awesome to get this merged soon, contrary to the package pr i can't release this as a plugin. And it will enable me and others to start using hyper v boxes exported from recent windows versions. 

Is there anything I can do to help more to get this going?

**Comment by b33pr** [comment]
I pulled these in and did a test - after I made the change in https://github.com/bbrala/vagrant/pull/2 everything worked. Thanks for all of your work on this

**Comment by bbrala** [comment]
Thank you for having a look. The condition seems fine like that indeed, good catch. Regarding bsdtar I'd like some extra info.

Does bsdtar give an error when you try to package? And what version of vagrant do you have installed? What version of windows? Did you use the gist as configuration? And you apparently merged both branches I made? Also how large is the hard disk that fails? 

(Sent by mobile)

**Comment by b33pr** [comment]
bsdtar was dumping an error 'the file seemed to change' but the packager
didn't notice. Running win 10 'anniversary edition' Did use the gist, I
think, did merge both branches, and absurdly large - like 13gb.

**Comment by bbrala** [comment]
@b33pr i tried 4 different vagrant hyper v boxes from hashicorp atlas. 

1. vagrant up
2. vagrant package
3. vagrant box add / init / up

Haven't run into any problems...

**Comment by DavidCzadilek** [comment]
@bbrala first off, thank you for your great work! build vagrant-dev according to your instructions and now importing hyper-v boxes works for the most part.

Ran into some furhter problems and i suppose it might have something to do with the fact that i was trying to use a 2nd generation hyper-v vm.
The Problem is, that no OS is detected on the VM. I've attached a screenshot with the difference in settings of the original and the vagrant-created vm in hyper-v-manager.
pasted image at 2016_10_19 02_40 pm

Seems to have something to do with setting the bootloader..
I'll be trying Gen1 for now.

Info:
* using vagrant-dev according to https://gist.github.com/bbrala/7e96be8a4e26f14834e8bbc21d246c38 & current master
* windows 10, hyper-v, all updates installed (not the win-insider ones)
* vm based on ubuntu server 14.04.5 x64 iso from ubuntu website
* hyper-v integration tools installed (by default), and added to boot-sequence, according to http://www.serverwatch.com/server-tutorials/installing-and-activating-hyper-v-linux-integration-services.html

**Comment by bbrala** [comment]
Thanks for having a look @DavidCzadilek. Looking at the imported VM it seems you packaged it with a custom XML which isn't needed anymore with this PR. Vagrant will look in the box file if there is a XML file and if so import it the old way. The old way doesn't support gen 2. You can do two things, remove the XML so there is only the binary config and repackage. Or also use the second PR i made to package it through vagrant. (https://github.com/mitchellh/vagrant/pull/7867).

**Comment by DavidCzadilek** [comment]
Thanks for the reply @bbrala!
I'm packaging manually according to https://www.vagrantup.com/docs/hyperv/boxes.html
My final box file (`*.tar.gz` -> `*.box`) holds the `Virtual Hard Disk` folder containing the hard drive `*.vhdx` file, the `Virtual Machines` folder containing a `*.vmcx` and a `*.VMRS` file, as well as a `metadata.json` with only the provider setting:
`{"provider":"hyperv"}`
Thats it - nothing else in the box.

**Comment by bbrala** [comment]
That is pretty weird. I'll have a go again and see what happens. It should copy the firmware...

**Comment by bbrala** [comment]
Did you tar on linux?

**Comment by DavidCzadilek** [comment]
Tried with both 7zip and Ubuntu bash on win10.
Havent worked much with powershell so far, especially scripting, but while patching I noticed that in the XML version of the import script there's a block for boot devices (https://github.com/mitchellh/vagrant/blob/master/plugins/providers/hyperv/scripts/import_vm.ps1#L86) which is missing entirely in the vmcx import script.

Tried to add it from the XML version, but that didnt seem to have an effect.. its possible i did something wrong though, since i have no experience with PS scripts (no debug errors though).

**Comment by bbrala** [comment]
Since the new script makes a copy from the machine itself, it copies the boot-state also. I've just exported a VM with the same image and settings as you and going to test if it will add it properly later tonight.

**Comment by bbrala** [comment]
Ok, i did the same as you described.

1. New Hyper V VM generation 2.
2. Install ubuntu server
3. Export 
4. tar with windows bash

Then i added the box to my vagrant install and loaded it up. It seems like it was a perfect copy with the bios in place.

importedboxes

I can use it without a problem... I did use this [article from Microsoft](
https://technet.microsoft.com/en-us/windows-server-docs/compute/hyper-v/supported-ubuntu-virtual-machines-on-hyper-v) to install the hyperv stuff though, it didn't report IP when i used your link (although i mightve missed something)

**Comment by DavidCzadilek** [comment]
Thanks for the test - the manual installation of services as described in the [article from Microsoft](https://technet.microsoft.com/en-us/windows-server-docs/compute/hyper-v/supported-ubuntu-virtual-machines-on-hyper-v) was exactly what I was missing. Everything up and running smoothly now! 👍

I'll be making another box based on Ubuntu 16.x using Packer now. Will keep you updated if I run into any  issues or find out more info worth sharing here.

**Comment by bbrala** [comment]
Awesome, glad it works as intended.

**Comment by bbrala** [comment]
Made two [Hyper V 2nd generation boxes](https://atlas.hashicorp.com/bbrala), built at work then repackaged at home with Vagrant without any issues.

**Comment by bbrala** [comment]
@chrisroberts it seems the PR for packer is slotted for the next release, which is 15 november. It would be a shame if this spawns new Hyper V generation 2 machines which do not work on Vagrant.

**Comment by taliesins** [comment]
@bbrala you can select the generation to use.

Whats the issue with generation 2 and vagrant? Is it simply that not all oses support generation 2?

**Comment by bbrala** [comment]
@taliesins BIggest problem with vagrant is that the new format for machines (vmcx) isn't supported, it cannot read that format. The wat it is fixed currectly, people use a custom XML with the configuration of the machine, but this means, no EFI bios is possible. 

So generation 2 is mostly broken because the default options (EFI bios) isn't supported at all.

**Comment by b33pr** [comment]
What's the status here?

**Comment by bbrala** [comment]
I'd really like to know if this realisticly will make it into 1.9 otherwise I will make it a plugin until release. But rather not do that if it will make it in.

**Comment by krasiyosifov** [comment]
I wanted to thank you @bbrala for your work! After fighting for two days with the problem of unable to import my own hyperv box into latest version of vagrant, I found your solution. 
To try your solution I had to install vagrant from source of your pull request. When I tested again to import the box, everything worked as expected with no errors! 
Now I'm happy with my setup and hoping that fix to be included in the official release.
If you need some assistance I can help you with testing various scenarios, because I want this fix to be released officially as soon as possible. I will leave the vagrant installation from the latest code from github pull request until it is officially released and working.
Thanks again!

File: plugins/providers/hyperv/scripts/import_vm_vmcx.ps1 [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+if ($generation -ne 1) {
+    Set-VMFirmware -VM $vmConfig.VM -EnableSecureBoot (Get-VMFirmware -VM $vmConfig.VM).SecureBoot
+}
+
+$report = Compare-VM -CompatibilityReport $vmConfig
+
+# Stop if there are incompatibilities
+if($report.Incompatibilities.Length -gt 0){
+    Write-Error-Message $(ConvertTo-Json $($report.Incompatibilities | Select -ExpandProperty Message))
+    exit 0
```
Comment (chrisroberts) [inline plugins/providers/hyperv/scripts/import_vm_vmcx.ps1]: Should this be returning back a non-zero status to indicate import failure?

Comment (bbrala) [inline plugins/providers/hyperv/scripts/import_vm_vmcx.ps1]: Hmm, sematicly, you are correct. But what i understood from how the code of Vagrant works it just parses the output and handles the error and prints the error in a nice way. Think the default error handling is a bit more rough in its messages. 

That is how i understood the code at least :)

Relevant code: https://github.com/bbrala/vagrant/blob/8744caebcf50095e808233fd7282849115eddacb/plugins/providers/hyperv/driver.rb#L20-L40

Comment (chrisroberts) [inline plugins/providers/hyperv/scripts/import_vm_vmcx.ps1]: :+1:

**Comment by bbrala** [comment]
Awesome, glad to see it slotted for 1.9 :)

**Comment by chrisroberts** [comment]
@bbrala I'm looking to have it included in 1.9.1

**Comment by bbrala** [comment]
@chrisroberts don't forget to close https://github.com/mitchellh/vagrant/issues/6102

**Comment by chrisroberts** [comment]
:+1:

**Pull request merged**